exporter/prometheusexporter/accumulator.go

```diff
@@ -172,8 +172,13 @@ func (a *lastValueAccumulator) accumulateGauge(metric pmetric.Metric, il pcommon
 func (a *lastValueAccumulator) accumulateSum(metric pmetric.Metric, il pcommon.InstrumentationScope, resourceAttrs pcommon.Map, now time.Time) (n int) {
 	doubleSum := metric.Sum()
 
-	// Drop metrics with non-cumulative aggregations
-	if doubleSum.AggregationTemporality() != pmetric.MetricAggregationTemporalityCumulative {
+	// Drop metrics with unspecified aggregations
+	if doubleSum.AggregationTemporality() == pmetric.MetricAggregationTemporalityUnspecified {
```

Comment:
Metrics like below are still being dropped by this condition:
`
# HELP solace_client_tx_bytes_total Number of transmitted bytes.
# TYPE solace_client_tx_bytes_total counter
solace_client_tx_bytes_total{client_name="#amqp/client/382b49dc52209f31/vert.x-ff7a82cb-28a7-4fe6-be41-ce65f07f01e8",client_username="womgr-sa-gke-nonprod",system="WOMGR",vpn_name="qa-gke-euwe4-cgeu-1"} 4.536923871e+09
`
Tested by applying this PR on top of `v0.51.0`

Reply:
Hi @artem-shorokhov, thanks for testing this out, could you add more details about the configuration/receiver you tested with? If it was the statsd, I doubt that you would need to enable the `is_monotonic_counter` option to set monotonic for the counter.

Reply:
Prometheus receiver scrapes Solace brokers via [Solace Prometheus Exporter](https://github.com/solacecommunity/solace-prometheus-exporter) which returns metrics in Prometheus format as shown above.

Here is the receiver config:
`
    prometheus/solace:
      config:
        scrape_configs:
        - job_name: client-stats
          metrics_path: /client-stats
          scrape_interval: 10s
          static_configs:
          - targets:
            - https://dev-aks-euwe-int-subdomain.messaging.solace.cloud:123
            ...
            - https://qa-gke-euwe4-cgeu-1-ts-1-int-subdomain.messaging.solace.cloud:123
            labels:
              systemName: eventm
              legalCompany: ingka
          relabel_configs:
          - source_labels: [__address__]
            target_label: __param_scrapeURI
          - target_label: __address__
            replacement: spe
          - source_labels: [__param_scrapeURI]
            target_label: instance
          metric_relabel_configs:
          - action: replace
            source_labels: [instance]
            regex: https://([a-z]+)-([a-z0-9]{3}).*
            replacement: $$2
            target_label: site
          - action: replace
            source_labels: [instance]
            regex: https://([a-z]+).*
            replacement: $$1
            target_label: environment
          - action: replace
            source_labels: [instance]
            regex: https://(.*)-.*-subdomain.*
            replacement: $$1
            target_label: service
`

Then metrics are grouped by the `system` label with `groupbyattrs` processor:
`
    groupbyattrs:
      keys:
      - system
`

And then routed to dedicated Prometheus exporters endpoints:
`
    routing:
      attribute_source: resource
      default_exporters:
      - logging
      from_attribute: system
      table:
      - value: SOLACE
        exporters:
        - prometheus/SOLACE
      ...
`

The pipeline looks like this:
`
      metrics/clients:
        receivers:
        - prometheus/solace
        processors:
        - memory_limiter
        - groupbyattrs
        - batch
        - routing
        exporters:
        - logging
        - prometheus/SOLACE
        ...
`

Reply:
could you add this file exporter in the pipeline as well:

```
exporters:
  file:
    path: ./metrics.json
```

It will generate the metric objects under JSON format. You could post it here as well.

I used to have an issue with the transform processor when it modified the metrics and left them in an error state after processing.

Reply:
For debugging purposes i've changed the pipeline a bit to export all the metrics from specific `system` to the `logging` and `prometheus` exporters at the same time:
```
    routing:
      attribute_source: resource
      default_exporters: []
      from_attribute: system
      table:
      - value: SOME_SYSTEM
        exporters:
        - prometheus/SOME_SYSTEM
        - logging
```

And also added logging to every condition in the `exporter/prometheusexporter/accumulator.go` to understand where precisely the metric is dropped.

So all the metrics were shown in logs, but all the counters were missing at the `prometheus` exporter endpoint.
And also logs showed that these metrics were just dropped by mentioned above condition.

Reply:
Yeah so we basically drop all the `pmetric.MetricAggregationTemporalityUnspecified` metric and non-monotonic delta metric. If you added some processors that alternate these metric attributes then they would got dropped as well. I guess the fix you need should be on the processor side.

Reply:
In my case only `counter` metrics were dropped by the `prometheus` exporter. `gauge` metrics were processed fine.

Reply:
Yes, this dropping logic is only applied for the counter.

> So all the metrics were shown in logs

Could you provide the metrics in the log exporter?

Reply:
Here are some metric that were in the logs but dropped by the `prometheus` exporter:
```
Metric #27
Descriptor:
     -> Name: solace_client_tx_bytes_total
     -> Description: Number of transmitted bytes.
     -> Unit: 
     -> DataType: Sum
     -> IsMonotonic: false
     -> AggregationTemporality: AGGREGATION_TEMPORALITY_UNSPECIFIED
NumberDataPoints #0
Data point attributes:
     -> client_name: STRING(#amqp/client/1b15b19052f77c86/vert.x-8a531b05-0750-4a64-b0a2-cf0b61bd6e19)
     -> client_username: STRING(username_here)
     -> environment: STRING(qa)
     -> legalCompany: STRING(company)
     -> service: STRING(service_name_here)
     -> site: STRING(gke)
     -> vpn_name: STRING(vpn_name_here)
StartTimestamp: 2022-05-20 17:55:04.127 +0000 UTC
Timestamp: 2022-05-20 17:55:04.127 +0000 UTC
Value: 0.000000
NumberDataPoints #1
Data point attributes:
     -> client_name: STRING(#amqp/client/2b613a6a219433ff/vert.x-2114d47b-1a43-488b-9e60-08871797cc20)
     -> client_username: STRING(username_here)
     -> environment: STRING(qa)
     -> legalCompany: STRING(company)
     -> service: STRING(service_name_here)
     -> site: STRING(gke)
     -> vpn_name: STRING(vpn_name_here)
StartTimestamp: 2022-05-20 17:55:04.127 +0000 UTC
Timestamp: 2022-05-20 17:55:04.127 +0000 UTC
Value: 0.000000
...
```

I assume it is dropped because `IsMonotonic: false`

Reply:
>  -> IsMonotonic: false
    -> AggregationTemporality: AGGREGATION_TEMPORALITY_UNSPECIFIED
    
 You have both of the dropping cases here. Refer to this link for more information: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/datamodel.md#sums-1
 
 I'm suspecting one of the processors is changing those attributes.

Reply:
Maybe. Will do some tests when I have some time and provide the results here.

Reply:
The culprit was `groupbyattrs` processor... and the fix is already on the `main`: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/9088/files